[rocSOLVER] Optionally use standard library for formatted printing rather than libfmt

john00003 opened the pull request:
## Motivation

Provide the option to use `std::format` rather than `fmt::format`. libfmt is not included by default on some operating systems.

## Technical Details

I've aliased all instances of `fmt::format` and `fmt::print` to be `rocsolver::format` and `rocsolver::print`. This alias will use either the standard library, or libfmt as a backend depending on CMake configuration.

CMake will automatically check if std::format is available to the user, and will use it unless explicitly overridden by the user with the CMake option `FORCE_USE_FMT_LIB_OPTION`. This is extended to the `install.sh` script as a command line option `--[no-]libfmt`

## Test Plan

Run rocSOLVER's logging tests in the GTest suite.

## Test Result

Tests pass when built with both the standard library and libfmt for fomatted printing.

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

john00003 committed:
aliased print, format, and join to rocsolver namespace functions

john00003 committed:
removed comments. removed unnecesary cmake options. some testings related to logging fail

john00003 committed:
moved C++ standard requirements/fmtlib requirements behind CMake option

john00003 committed:
added fmt wrapper required when aliasing. formatted some files with clang-format. build working for both fmt and std backend

john00003 committed:
fixed seperator character being added at end of joined tuples

john00003 committed:
modified namespace usage

john00003 committed:
refactored extra formatter type

john00003 committed:
tidied CMakeLists.txt

john00003 committed:
formatted files

john00003 committed:
removed linker arguments (unnecessary unless libstdc++ is not identified with  or

john00003 committed:
made USE_FMT_LIB default ON. updated install.sh with an option to not use fmt lib

john00003 committed:
changed required fmt version to original - no need for newer version

john00003 committed:
cleanup

john00003 committed:
cleanup

assistant-librarian[bot] committed:
Merge commit '3bdee8caa5d8458d7c4a6831530458354cc293bc' into import/develop/john00003_rocSolver/fmt_lib

john00003 committed:
Merge branch 'develop' into import/develop/john00003_rocSolver/fmt_lib

john00003 committed:
updated rocsolver aliased print to use std::format rather than std::print to lower C++ standard requirements

john00003 committed:
added new bash command line options to force option to use or not use libfmt. changed check for std::format to check for existence of C++ standard header file

john00003 committed:
fixed install script to add CMake commands to correct variable name

john00003 committed:
applied formatting

john00003 committed:
removed libstdc++ CMake option. any standard library that allows for std::format can be used

john00003 committed:
CI tests bugged. retrigger

mgates3 commented on projects/rocsolver/common/include/fmt_rocblas_types.hpp:
```diff
@@ -27,20 +27,34 @@
+    {
+#define END_CONDITIONAL_NAMESPACE }
+#define ROCSOLVER_FMT_CONST const
+#endif
```
I guess this is just how our clang-format style is, but personally, I would prefer to indent all the `#if` stuff the same way we indent regular `if` stuff. (I may be in a minority among C programmers here.) I.e.,
```
#ifdef USE_FMT_LIB
    #include <fmt/format.h>
    #include <fmt/ostream.h>
    #define BEGIN_CONDITIONAL_NAMESPACE \
        namespace fmt                   \
        {
    #define END_CONDITIONAL_NAMESPACE }
    #if FMT_VERSION < 80000
        #define ROCSOLVER_FMT_CONST
    #else
        #define ROCSOLVER_FMT_CONST const
    #endif
#else
    #include <format>
    #define BEGIN_CONDITIONAL_NAMESPACE \
        namespace std                   \
        {
    #define END_CONDITIONAL_NAMESPACE }
    #define ROCSOLVER_FMT_CONST const
#endif
```

john00003 committed:
removed unconditional debug logging

john00003 committed:
renamed c++ variable to cxx to match CMake conventions

mgates3 commented on projects/rocsolver/common/include/rocsolver_utility.hpp:
```diff
@@ -0,0 +1,167 @@
+        auto it = ctx.begin();
+        auto end = ctx.end();
+
+        if(it != ctx.end() && *it != '}')
```
Minor: you already got `end = ctx.end`, why call `ctx.end` again?

john00003 replied:
This is addressed!

mgates3 commented on projects/rocsolver/common/include/rocsolver_utility.hpp:
```diff
@@ -0,0 +1,167 @@
+        auto it = ctx.begin();
+        auto end = ctx.end();
+
+        if(it != ctx.end() && *it != '}')
```
Same, calling `ctx.end` twice.

mgates3 commented on projects/rocsolver/common/include/rocsolver_utility.hpp:
```diff
@@ -0,0 +1,167 @@
+    template <class... Args>
+    void print(std::ostream& stream, std::format_string<Args...> fmt_string, Args&&... args)
+    {
+        printf("printed with std::print (actually format) \n");
```
Was this `printf` for debugging?

john00003 replied:
Yes, shameful! This is addressed :)

mgates3 commented on projects/rocsolver/common/include/rocsolver_utility.hpp:
```diff
@@ -0,0 +1,167 @@
+#else
+        joinable_range range{r, sep};
+        return std::format("{}", range);
+#endif
```
Again, this is probably just how clang-format & our style are, but I personally indent `#if` things with the code. I'm definitely in the minority here, but I think the code is easier to read. I.e.,
```
    std::string join(Range&& r, std::string_view sep)
    {
        #ifdef USE_FMT_LIB
            return fmt::format("{}", fmt::join(r, sep));
        #else
            joinable_range range{r, sep};
            return std::format("{}", range);
        #endif
    }
```

john00003 replied:
I believe that this is a disallowed by our clang-format guidelines.

However, I like your opinion, and I am on the same page!

mgates3 commented on projects/rocsolver/common/include/rocsolver_utility.hpp:
```diff
@@ -0,0 +1,167 @@
+        return std::format_to(ctx.out(), "{}", result);
+    }
+};
+#endif
```
I don't know if it's a convention in rocSolver, but I find it helpful to comment what the `#endif` refers to if they are widely separated.
```suggestion
#endif  // not USE_FMT_LIB
```

mgates3 commented on projects/rocsolver/common/include/rocsolver_utility.hpp:
```diff
@@ -0,0 +1,167 @@
+    {
+        return std::format(fmt_string, std::forward<Args>(args)...);
+    }
+#else
```
```suggestion
#else  // USE_FMT_LIB
```

mgates3 commented on projects/rocsolver/common/include/rocsolver_utility.hpp:
```diff
@@ -0,0 +1,167 @@
+    {
+        return fmt::format(std::forward<T>(fmt_string), std::forward<Args>(args)...);
+    }
+#endif
```
```suggestion
#endif  // USE_FMT_LIB
```

mgates3 commented on projects/rocsolver/common/CMakeLists.txt:
```diff
@@ -27,6 +27,22 @@
 
 add_library(rocsolver-common INTERFACE)
 
+# add cmake 23 requirement
```
Do you mean C++ 23, not "cmake 23"?

john00003 replied:
This is addressed!

mgates3 commented on projects/rocsolver/library/src/include/rocsolver_logvalue.hpp:
```diff
@@ -27,12 +27,16 @@
+#define LIB_NAMESPACE fmt
+#else
+#define LIB_NAMESPACE std
+#endif
```
Or a macro with `ROCSOLVER_` or similar prefix.

john00003 replied:
This is addressed!

mgates3 commented on projects/rocsolver/CMakeLists.txt:
```diff
@@ -66,9 +65,35 @@ option(ROCSOLVER_USE_REFERENCE_SECULAR_EQUATIONS_SOLVER "Use LAPACK like impleme
+option(FORCE_USE_FMT_LIB_OPTION "Use the user provided USE_FMT_LIB option, rather than the result of the automated check by CMake if std::format is available" OFF)
+option(USE_FMT_LIB "Use the fmt library instead of std::print" OFF)
+message(STATUS "Value of FORCE_USE_FMT_LIB: ${FORCE_USE_FMT_LIB_OPTION}")
+message(STATUS "Value of USE_FMT_LIB: ${USE_FMT_LIB}")
```
I'm wondering if this can be encoded in a single option, rather than passing 2 flags. E.g.,
```
cmake ...                    # USE_FMT_LIB default value
cmake -DUSE_FMT_LIB=on  ...  # USE_FMT_LIB forced on
cmake -DUSE_FMT_LIB=off ...  # USE_FMT_LIB forced off
```
Perhaps by setting it to "auto" here:
```
option( USE_FMT_LIB "..." "auto" )
```
and below checking:
```
if (USE_FMT_LIB STREQUAL "auto")
    # check std::format support
    ...
    if (CXX_FORMAT_SUPPORT STREQUAL "")
        set(USE_FMT_LIB ON)
    else()
        set(USE_FMT_LIB OFF)
    endif()
endif()
```

john00003 replied:
This is addressed! Good idea, much cleaner solution :)

mgates3 commented on projects/rocsolver/install.sh:
```diff
@@ -615,6 +627,16 @@ if [[ "${build_codecoverage}" == true ]]; then
+        cmake_common_options+=('-DFORCE_USE_FMT_LIB_OPTION=ON')
+        cmake_common_options+=('-DUSE_FMT_LIB=OFF')
+    fi
+fi
```
See my comment on projects/rocsolver/CMakeLists.txt. Similarly here, could we remove `force_libfmt`, set `libfmt=auto` by default, and use:
```
if [[ "${libfmt}" == true ]]; then
    cmake_common_options+=('-DUSE_FMT_LIB=ON')
else if [[ "${libfmt}" == false ]]; then
    cmake_common_options+=('-DUSE_FMT_LIB=OFF')
fi
```

john00003 replied:
This is addressed!

mgates3 reviewed (changes requested):
You can ignore the `#if` indentation comments. They're just my opinion.

john00003 committed:
added tests for rocsolver formatting

john00003 committed:
added compile check to determine whether to use std or fmt for formatted printing

john00003 committed:
renamed macros with ROCSOLVER in the title to avoid collisions and global leakage

john00003 committed:
remove redundant operations in parse functions. adjust error report for cleaner multi-line errors. use std::basic_format_parse_context

basic_format_pase_context no longer assumed char template instantiation,
and does not allow for wchar

john00003 committed:
fixed typo in CMake comment checking for c++23

john00003 committed:
Add comment for end of #ifdef

Co-authored-by: Mark Gates <mark.gates@amd.com>

john00003 committed:
Add comment for end of #ifdef

Co-authored-by: Mark Gates <mark.gates@amd.com>

john00003 committed:
Add comment for end of #ifdef

Co-authored-by: Mark Gates <mark.gates@amd.com>

john00003 commented on projects/rocsolver/CMakeLists.txt:
```diff
@@ -66,9 +65,35 @@ option(ROCSOLVER_USE_REFERENCE_SECULAR_EQUATIONS_SOLVER "Use LAPACK like impleme
+    else()
+        set(USE_FMT_LIB OFF)
+    endif()
+endif()
```
This is addressed!

john00003 commented on projects/rocsolver/common/include/rocsolver_utility.hpp:
```diff
@@ -0,0 +1,167 @@
+        if(it != ctx.end() && *it != '}')
+            throw std::format_error(
+                "Invalid format args provided for rocsolver::formatting::join. The format string '{}' must be used. \n \
+					The joinable_range type should be privately used by rocsolver::formatting::join.");
```
This is addressed!

john00003 commented on projects/rocsolver/common/include/rocsolver_utility.hpp:
```diff
@@ -0,0 +1,167 @@
+    }
+};
+
+template <typename... T, typename CharT>
+struct std::formatter<joinable_range<std::tuple<T...>>, CharT>
+{
```
This is addressed! I've added some tests.

john00003 committed:
changed FORCE_USE_FMT_LIB_OPTION to just a single CMake variable.

this variable is cached between runs, and when unspecified will attempt to detect if std::format and std::print are available on the user's system.

john00003 committed:
Merge branch 'users/john00003/rocsolver-std-format' of https://github.com/john00003/rocm-libraries into users/john00003/rocsolver-std-format

john00003 committed:
Merge branch 'develop' into users/john00003/rocsolver-std-format

john00003 committed:
apply formatting to double space after #endif

john00003 committed:
Merge branch 'develop' into users/john00003/rocsolver-std-format

jzuniga-amd commented:
I think it's a good idea to remove the fmt dependency if possible. 
I have no technical comments on this. Leaving reviews/approvals to others.

john00003 committed:
Merge branch 'develop' into users/john00003/rocsolver-std-format

Ozone-Whiskey commented on projects/rocsolver/common/include/rocsolver_utility.hpp:
```diff
@@ -0,0 +1,163 @@
+#include <fmt/ranges.h>
+#else
+#include <cstdio>
+#include <format>
+#include <iostream>
+#include <ostream>
```
Added after consulting John

Ozone-Whiskey commented on projects/rocsolver/CMakeLists.txt:
```diff
@@ -66,9 +65,43 @@ option(ROCSOLVER_USE_REFERENCE_SECULAR_EQUATIONS_SOLVER "Use LAPACK like impleme
+# Configure which formatting dependency to use
+
+option(ROCSOLVER_EMBED_FMT "Hide libfmt symbols" ON)
+
+set(USE_FMT_LIB "auto" CACHE STRING "Use the fmt library instead of std::format (auto, ON, or OFF)")
+
+if(USE_FMT_LIB STREQUAL "auto")
+    # Verify std::format and std::print actually compile, not just that headers exist
+    include(CheckCXXSourceCompiles)
+    set(CMAKE_REQUIRED_FLAGS "-std=c++23")
+    check_cxx_source_compiles("
```
Added

Ozone-Whiskey commented on projects/rocsolver/CMakeLists.txt:
```diff
@@ -66,9 +65,43 @@ option(ROCSOLVER_USE_REFERENCE_SECULAR_EQUATIONS_SOLVER "Use LAPACK like impleme
+            return 0;
+        }
+    " CXX_FORMAT_SUPPORT)
+    message(STATUS "C++ std::format/std::print support: ${CXX_FORMAT_SUPPORT}")
```
Added this as a fallback.

mgates3 commented on projects/rocsolver/clients/gtest/logging_gtest.cpp:
```diff
@@ -436,3 +439,30 @@ TEST_F(checkin_misc_LOGGING, end_before_begin)
+{
+    int a = 1;
+    double b = 2.5;
+    auto t = std::tie(a, b);
```
Is there a reason to use `std::tie` instead of `std::tuple` here?

I had not seen std::tie before, so that's interesting to learn about. I gather that it's usually used on the left hand side to unpack a tuple, e.g.,
```
double a, b;
std::tie( a, b ) = std::tuple( 1.0, 2.0 );
```

Ozone-Whiskey replied:
Yeah it was interesting to learn about too. I just inherited it from John but either way should work. It does seem to be convention that tie is used in LHS. 

Let me know if you would like me to change that to be more conventional.

EdDAzevedo commented on projects/rocsolver/clients/common/misc/client_environment_helpers.cpp:
```diff
@@ -1,5 +1,5 @@
 /* **************************************************************************
- * Copyright (C) 2022-2024 Advanced Micro Devices, Inc. All rights reserved.
+ * Copyright (C) 2022-2025 Advanced Micro Devices, Inc. All rights reserved.
```
should it be 2026?

EdDAzevedo commented on projects/rocsolver/clients/common/misc/clients_utility.cpp:
```diff
@@ -1,5 +1,5 @@
 /* **************************************************************************
- * Copyright (C) 2018-2024 Advanced Micro Devices, Inc. All rights reserved.
+ * Copyright (C) 2018-2025 Advanced Micro Devices, Inc. All rights reserved.
```
should it be 2026?

Ozone-Whiskey replied:
fixed

EdDAzevedo commented on projects/rocsolver/clients/gtest/rocsolver_gtest_main.cpp:
```diff
@@ -1,5 +1,5 @@
 /* **************************************************************************
- * Copyright (C) 2020-2022 Advanced Micro Devices, Inc. All rights reserved.
+ * Copyright (C) 2020-2025 Advanced Micro Devices, Inc. All rights reserved.
```
should it be 2026?

Ozone-Whiskey replied:
Yes, I will update that, thanks!

Ozone-Whiskey commented on projects/rocsolver/CMakeLists.txt:
```diff
@@ -66,9 +65,52 @@ option(ROCSOLVER_USE_REFERENCE_SECULAR_EQUATIONS_SOLVER "Use LAPACK like impleme
+
+option(ROCSOLVER_EMBED_FMT "Hide libfmt symbols" ON)
+
+set(USE_FMT_LIB "ON" CACHE STRING "Use the fmt library instead of std::format (auto, ON, or OFF)")
```
As per conversation with Julio, auto being default is actually undesirable behavior, and we want the user to explicitly set this to auto as mismatched versioning on print from rest of the user's C++ version could falsely trigger the switch.
